Allow Sites to use WP_POST _REVISIONS for other post types

NeilWJames opened the pull request:
This plugin really requires sites to not limit the number of revisions. But sites may wish to restrict the number for other post types.
This change, by default, ensures that revisions are unlimited for documents.
As is common practice, a filter has been provided to set a specific number for documents.
Clearly 0 should be considered an error, and messages will be produced when editing existing documents where the number is at or beyond the user defined limit as making changes will cause revision deletion.

benbalter commented on includes/class-wp-document-revisions-admin.php:
```diff
@@ -884,6 +890,83 @@ public function profile_update_cb() {
+	 */
+	public function manage_document_revisions( $num, $post ) {
+
+		if ( ! is_object( $post ) || 'document' !== $post->post_type ) {
```
Would it make sense to use `verify_post_type` here?

NeilWJames replied:
Agreed

benbalter commented on includes/class-wp-document-revisions-admin.php:
```diff
@@ -884,6 +890,83 @@ public function profile_update_cb() {
+
+		global $post;
+
+		if ( ! is_object( $post ) || 'document' !== $post->post_type ) {
```
Would it make sense to use `verify_post_type` here?

NeilWJames replied:
Agreed

benbalter commented on includes/class-wp-document-revisions-admin.php:
```diff
@@ -884,6 +890,83 @@ public function profile_update_cb() {
+			// need to check that we're not at the limit
+			$revisions = count( wp_get_post_revisions( $post->ID ) );
+
+			if ( $num < $revisions ) {
```
Are the conditionals and messages switched here? I'd think `num == revision` would be maximum revisions reached, and and `num < revisions` to be more revisions exist than permitted?

NeilWJames replied:
Thank you for your review. I'll make the changes once this is sorted.

Once implemented (with a user defined limit), and only when the user makes a change will any existing version be deleted. 

So once the user has `num` versions, then any change will leave the user with `num` versions.

But if there already exists more than `num` versions , it only displays `num` versions but the others are still there - until an update is made when they will be removed.

So this can occur when the limit is changed (or first implemented).

At one level, an update will delete existing data, so they could be combined into a single `$num <= $version`  warning

benbalter commented on includes/class-wp-document-revisions-admin.php:
```diff
@@ -884,6 +890,83 @@ public function profile_update_cb() {
+		if ( 0 === $num ) {
+			// setting revisions to 0 makes no sense for this plugin
+			// @codingStandardsIgnoreLine WordPress.XSS.EscapeOutput.OutputNotEscaped
+			echo '<div class="notice notice-error"><p>';
```
Would it make sense to `?><div class="notice notice-error"><p><?php` to avoid having to disable the linter?

NeilWJames replied:
Agreed. It will remove clutter.

NeilWJames commented on includes/class-wp-document-revisions-admin.php:
```diff
@@ -884,6 +890,83 @@ public function profile_update_cb() {
+	 * @param int     $num  default value for the number of revisions for the post_type
+	 * @param WP_Post $post current post
+	 */
+	public function manage_document_revisions( $num, $post ) {
```
Probably better be manage_document_revisons_limit. 
You are correct that it does not manage the revisions themselves and its current name would indicate that.

benbalter commented on includes/class-wp-document-revisions-admin.php:
```diff
@@ -1355,8 +1456,34 @@ public function delete_attachments_with_document( $post_id ) {
+
+				// make sure that the attachment is not refered to by another post
+				global $wpdb;
+				$doc_link = $wpdb->get_var( $wpdb->prepare( "SELECT COUNT(1) FROM $wpdb->posts WHERE %d IN (post_parent, id) AND post_content = %d AND post_name NOT LIKE '%d-autosave-v1' ",
```
Fixed the PHPCS errors I could. Based on https://wordpress.stackexchange.com/questions/8825/how-do-you-properly-prepare-a-like-sql-statement and https://www.itsupportguides.com/knowledge-base/wordpress/wordpress-how-to-use-like-in-wpdb-prepare-sql-statements/, I think this should be:

```php
$doc_link = $wpdb->get_var(
					$wpdb->prepare(
						"SELECT COUNT(1) FROM $wpdb->posts WHERE %d IN (post_parent, id) AND post_content = %d AND post_name NOT LIKE %s",
						$document->post_parent,
						$document->post_content,
						$wpdb->esc_like( $document->post_parent . '-autosave-v1' )
					)
				);
```

If that still achieves the desired outcome?

NeilWJames commented:
Sorry for ignoring your comment - I was too busy "fighting" SourceTree and the continually changing phpcs rules to notice this.
Thanks for pointing out how to do the "like" commands in a prepare statement. As you see I changed it to a straight !=.
Whilst testing with a specific number of revisions, I did come across a similar issue to the one with "serve_file" in that the attachment gets deleted using a cached standard directory and not the document directory. So I added an extra deletion using the document directoy in that case.

NeilWJames commented:
Oops. Thought I was closing the conversation at the lower level, not the pull request.
